lms/djangoapps/course_api/blocks/views.py

```diff
@@ -303,3 +303,93 @@ def list(self, request, hide_access_denials=False):  # pylint: disable=arguments
+
+        Returns the blocks in the course according to the requesting user's
+        access level.
+
+    **Example requests**:
+
+        GET /api/courses/v1/blocks/?course_id=<course_id>
+        GET /api/courses/v1/blocks/?course_id=<course_id>
+            &username=anjali
+            &depth=all
+            &requested_fields=graded,format,student_view_multi_device,lti_url
+            &block_counts=video
+            &student_view_data=video
+            &block_types_filter=problem,html
+
+    **Parameters**:
+
+        This view redirects to /api/courses/v2/blocks/<root_usage_key>/ for the
+        root usage key of the course specified by course_id.  The view accepts
+        all parameters accepted by :class:`BlocksInCourseView`
+
+    **Response Values**
+
+        Responses are identical to those returned by :class:`BlocksInCourseView` when
+        passed the root_usage_key of the requested course but there is one exception explained below.
+
+        It adds completion from root of the course to every child block and make it boolean i.e
+        completion will be either 1 or 0
+
+    """
+
+    def list(self, request, hide_access_denials=False):
+        """
+        Retrieves the usage_key for the requested course, and then returns the
+        same information that would be returned by BlocksView.list, called with
+        that usage key
+
+        Arguments:
+            request - Django request object
+        """
+
+        response = super().list(request, hide_access_denials=hide_access_denials)
+
+        course_blocks = {}
+        if request.query_params.get('return_type') == 'list':
+            for course_block in response.data:
+                course_blocks[course_block['id']] = course_block
+
+                if course_block.get('type') == 'course':
+                    root = course_block['id']
+        else:
+            root = response.data['root']
+            course_blocks = response.data['blocks']
+
+        recurse_mark_complete(root, course_blocks)
+
+        return response
+
+
+def recurse_mark_complete(block_id, blocks):
+    """
+    Helper function to walk course tree dict,
+    marking completion as 1 or 0
+
+    If all blocks are complete, mark parent block complete
+
+    :param blocks: dict of all blocks
+    :param block_id: root or child block id
+
+    :return:
+        block: course_outline_root_block block object or child block
+    """
+    block = blocks.get(block_id, {})
+    if block.get('completion') == 1:
+        return
+
+    child_blocks = block.get('children', block.get('descendents'))
+    if child_blocks:
+        for child_block in child_blocks:
+            recurse_mark_complete(child_block, blocks)
+
+        completable_blocks = [blocks[child_block_id] for child_block_id in child_blocks
+                              if blocks[child_block_id].get('type') != 'discussion']
+        block['completion'] = int(all(child.get('completion') == 1 for child in completable_blocks))
```

Comment:
why specifically exclude discussion blocks? relatedly, will this handle ORAs, staff-graded assignments, and any other random block type?

Reply:
I am trying to bring web feature into mobile for completion marks, therefor I copied their logic for calculating completion. Here is the[ reference ](https://github.com/edx/edx-platform/blob/master/openedx/features/course_experience/utils.py#L112)